sdk/go/common/workspace/plugins.go

```diff
@@ -317,14 +318,148 @@ func (info PluginInfo) Download() (io.ReadCloser, int64, error) {
+	}{}
+	err = json.Unmarshal(jsonBody, &release)
+	if err != nil {
+		logging.V(9).Infof("github json response: %s", jsonBody)
+		logging.V(9).Infof("cannot unmarshal github response: %s", err.Error())
+		return nil, -1, err
+	}
+	assetURL := ""
+	for _, asset := range release.Assets {
+		if asset.Name == assetName {
+			assetURL = asset.URL
+		}
+	}
+	if assetURL == "" {
+		logging.V(9).Infof("github json response: %s", jsonBody)
+		logging.V(9).Infof("plugin asset '%s' not found", assetName)
+		return nil, -1, errors.Errorf("plugin asset '%s' not found", assetName)
+	}
+	if ghToken, ok := os.LookupEnv("GITHUB_TOKEN"); ok {
+		return authenticatedGetPluginResponse(assetURL, "token", ghToken, "", "")
+	} else if paToken, ok := os.LookupEnv("GITHUB_PERSONAL_ACCESS_TOKEN"); ok {
+		if actor, ok := os.LookupEnv("GITHUB_ACTOR"); ok {
+			return authenticatedGetPluginResponse(assetURL, "", "", paToken, actor)
+		}
+	}
+	return nil, -1, errors.New("no authentication information provided")
+}
+
+func getPluginPrivateGitHubReleaseAPIResponse(info PluginInfo) (io.ReadCloser, int64, error) {
+	repoOwner := os.Getenv("GITHUB_REPOSITORY_OWNER")
+	if repoOwner == "" {
+		return nil, -1, errors.New("ENV[GITHUB_REPOSITORY_OWNER] not set")
+	}
+	ghToken := os.Getenv("GITHUB_TOKEN")
+	paToken := ""
+	actor := ""
+	if ghToken == "" {
+		paToken = os.Getenv("GITHUB_PERSONAL_ACCESS_TOKEN")
+		actor = os.Getenv("GITHUB_ACTOR")
+	}
+	if ghToken == "" && (paToken == "" || actor == "") {
+		return nil, -1, errors.New("no GitHub authentication information provided")
+	}
+	releaseURL := fmt.Sprintf("https://api.github.com/repos/%s/pulumi-%s/releases/tags/v%s",
+		repoOwner, info.Name, info.Version.String())
+	logging.V(9).Infof("plugin GitHub releases url: %s", releaseURL)
+
+	req, err := http.NewRequest("GET", releaseURL, nil)
+	if err != nil {
+		return nil, -1, err
+	}
+
+	userAgent := fmt.Sprintf("pulumi-cli/1 (%s; %s)", version.Version, runtime.GOOS)
+	req.Header.Set("User-Agent", userAgent)
+	// the OAuth token takes precedence
+	if ghToken != "" {
+		req.Header.Set("Authentication", fmt.Sprintf("token %s", ghToken))
+	} else if paToken != "" && actor != "" {
+		req.SetBasicAuth(actor, paToken)
+	}
+
+	logging.V(9).Infof("plugin install request headers: %v", req.Header)
+
+	resp, err := httputil.DoWithRetry(req, http.DefaultClient)
+	if err != nil {
+		return nil, -1, err
+	}
+
+	logging.V(9).Infof("plugin install response headers: %v", resp.Header)
+
+	if resp.StatusCode < 200 || resp.StatusCode > 299 {
+		logging.V(9).Infof("%d HTTP error fetching plugin from %s", resp.StatusCode, releaseURL)
+		return nil, -1, errors.Errorf("%d HTTP error fetching plugin GitHub releases from %s", resp.StatusCode, releaseURL)
+	}
+
+	return resp.Body, resp.ContentLength, nil
+}
+
+func authenticatedGetPluginResponse(assetURL, tokenType, token, secret, username string) (io.ReadCloser, int64, error) {
+	logging.V(9).Infof("full plugin download url: %s", assetURL)
```

Comment:
The amount of general code duplication in this file is a bit sad but maybe that's something for us to cleanup after.

Reply:
Following up on this conversation, I did some refactoring to reduce duplication.